File: quant1x/contrib/calendar.py
# -*- coding: utf-8 -*-
# Copyright (c) Quant1X <wangfengxy@sina.cn>.
# Licensed under the MIT License.

import os
import time
from turtle import up
#from functools import lru_cache
import numpy as np
import pandas as pd
from typing import List, Optional
import bisect
import requests
import csv
from .sina.decoder import FinanceDecoder
from quant1x.runtime.once import RollingOnce
from quant1x.data import market, layout, cache, config
from quant1x.data.timestamp import Timestamp
from quant1x.std import filesystem as fs
from quant1x.log import logger


# 新浪财经交易日历URL
SINA_CALENDAR_URL = "https://finance.sina.com.cn/realstock/company/klc_td_sh.txt"
CALENDAR_MISSING_DATE = "1992-05-04"


# in-memory caches (match Go implementation)
globalCalendarsString = []
globalCalendarsTimestamp = []

calendarRollingOnce = RollingOnce(name='calendar_init', cron=market.cn_cron_expr_daily_init)

def get_calendar_filename() -> str:
    return os.path.join(config.meta_path, "calendar")

def get_calendar_marker_filename() -> str:
    return os.path.join(config.meta_path, "calendar.updated")


def __preprocess(text: str) -> str:
    """预处理JS-like响应文本（去除赋值、尾部分号和引号）"""
    s = text
    if "=" in s:
        pos = s.find("=")
        s = s[pos + 1:]
    if ";" in s:
        pos = s.find(";")
        s = s[:pos]
    s = s.replace('"', '')
    return s

def __decode(text: str) -> List[str]:
    """解码日历数据"""
    pre = __preprocess(text)
    decoder = FinanceDecoder(pre)
    raw = decoder.decode()

    dates = []
    if isinstance(raw, list):
        for item in raw:
            if isinstance(item, dict) and "date" in item:
                date = item["date"].strip()
                if date:
                    dates.append(date)
            elif isinstance(item, str):
                dates.append(item.strip())

    return dates if dates else []

def update_calendar():
    """下载交易日历数据并缓存到磁盘"""
    fn = get_calendar_filename()
    logger.warning(f"更新交易日历: 下载并缓存到 {fn}")
    try:
        # conditional GET using file mtime to avoid unnecessary downloads
        headers = {}
        if os.path.exists(fn):
            try:
                mtime = os.path.getmtime(fn)
                # format in RFC1123
                from email.utils import formatdate

                headers['If-Modified-Since'] = formatdate(mtime, usegmt=True)
            except Exception:
                pass

        resp = requests.get(SINA_CALENDAR_URL, timeout=15, headers=headers)
        if resp.status_code == 304:
            # 未修改，直接返回
            return
        resp.raise_for_status()

        body = resp.text
        dates = __decode(body)
        if not dates:
            raise ValueError("解码交易日历数据失败")

        # Ensure missing date present, insert preserving order (do not resort)
        idx = bisect.bisect_left(dates, CALENDAR_MISSING_DATE)
        if idx == len(dates) or dates[idx] != CALENDAR_MISSING_DATE:
            dates.insert(idx, CALENDAR_MISSING_DATE)

        # write CSV cache
        os.makedirs(os.path.dirname(fn), exist_ok=True)
        with open(fn, 'w', newline='', encoding='utf-8') as f:
            writer = csv.writer(f)
            writer.writerow(["date", "source"])
            for date in dates:
                writer.writerow([date, "sina"])

        # set file mtime: prefer HTTP Last-Modified when present; fall back to now.
        now_ts = time.time()
        lm = resp.headers.get('Last-Modified')
        if lm:
            try:
                from email.utils import parsedate_to_datetime

                dt = parsedate_to_datetime(lm)
                secs = int(dt.timestamp())
                os.utime(fn, (secs, secs))
            except Exception:
                try:
                    secs = int(now_ts)
                    os.utime(fn, (secs, secs))
                except Exception:
                    pass
        else:
            try:
                secs = int(now_ts)
                os.utime(fn, (secs, secs))
            except Exception:
                pass

        # update the calendar marker next to the cache so RollingOnce and
        # other processes can decide when the last successful update occurred.
        try:
            marker = get_calendar_marker_filename()
            os.makedirs(os.path.dirname(marker), exist_ok=True)
            # create or truncate marker file
            fs.update_file_mtime(marker, now_ts)
        except Exception:
            pass
    except Exception as e:
        # match Go behaviour: let caller decide; raise to caller
        raise RuntimeError(f"更新交易日历失败: {e}")


def lazy_load_calendar():
    marker = get_calendar_marker_filename()
    now_time = Timestamp.now()
    mod_time = cache.get_filename_modified_time(marker)
    today_init_time = now_time.get_pre_market_time()
    if now_time > today_init_time and mod_time < today_init_time:
        logger.debug("交易日历缓存文件过期，执行更新")
        try:
            update_calendar()
            fs.update_file_mtime(marker)
        except Exception:
            pass
    else:
        logger.debug("交易日历缓存文件未过期，跳过更新")
    logger.debug("加载交易日历缓存文件到内存")
    #ensure_updated = False
    fn = get_calendar_filename()
    if not os.path.exists(fn):
        logger.debug("交易日历缓存文件不存在，跳过加载")
        return

    try:
        with open(fn, newline='', encoding='utf-8') as f:
            reader = csv.reader(f)
            # skip header
            try:
                next(reader)
            except StopIteration:
                return
            ss = []
            ts = []
            for rec in reader:
                if not rec:
                    continue
                date = rec[0].strip()
                if not date:
                    continue
                ss.append(date)
                try:
                    t = Timestamp.parse(date)
                    ts.append(t.get_pre_market_time())
                except Exception:
                    pass
    except Exception:
        logger.error("加载交易日历缓存文件失败，跳过加载")
        return

    global globalCalendarsString, globalCalendarsTimestamp
    globalCalendarsString = ss
    globalCalendarsTimestamp = ts

#@lru_cache(maxsize=None)
def load_calendar() -> pd.Series:
    """
    交易日历
    """
    # Ensure in-memory cache is loaded using the persistent rolling once (do -> swallow exceptions)
    calendarRollingOnce.do(lazy_load_calendar)
    if not globalCalendarsString:
        raise RuntimeError("exchange calendar is empty")
    # return a pandas Series for compatibility
    return pd.Series(globalCalendarsString)

#@lru_cache(maxsize=None)
def __calendar_timestamps() -> List[Timestamp]:
    """
    交易日历 (Timestamp对象列表)
    """
    dates = load_calendar()
    # 转换为 Timestamp 对象，并设置为当天的盘前时间 (09:00:00)
    return [Timestamp.parse(d).get_pre_market_time() for d in dates]

def calendar() -> pd.Series:
    """
    获取全部的交易日期
    Returns:
        pd.Series
    """
    return load_calendar()


def fix_trade_date(date_str: str, fmt: str = "%Y-%m-%d") -> str:
    """强制将日期字符串转换为指定格式

    参数:
        date_str: 输入日期字符串
        fmt: 目标格式（默认%Y-%m-%d）

    返回:
        统一格式的日期字符串

    示例:
        >>> fix_trade_date("2023/12/25")
        "2023-12-25"
    """
    from datetime import datetime
    return datetime.strptime(date_str, "%Y-%m-%d").strftime(fmt) if date_str else date_str


def get_today() -> str:
    """
    获取当前日期
    """
    date = time.strftime(layout.FORMAT_ONLY_DATE)
    return date

def last_trading_day(date: Optional[Timestamp] = None, debug_timestamp: Optional[Timestamp] = None) -> Timestamp:
    """
    获取最近一个交易日 (Timestamp版本)
    """
    trade_dates = __calendar_timestamps()
    if not trade_dates:
        return Timestamp.zero()

    if date is None:
        # 默认使用今天
        date = Timestamp.now().get_pre_market_time()

    # 查找 date 在 trade_dates 中的位置 (upper_bound)
    # bisect_right 相当于 C++ 的 upper_bound
    idx = bisect.bisect_right(trade_dates, date)
    
    if idx > 0:
        idx -= 1
    
    # 判断是否盘前
    last_ts = trade_dates[idx]
    current_ts = debug_timestamp if debug_timestamp is not None else Timestamp.now()
    
    if current_ts < last_ts and idx > 0:
        idx -= 1
        
    return trade_dates[idx]


def prev_trading_day(date: Optional[Timestamp] = None, debug_timestamp: Optional[Timestamp] = None) -> Timestamp:
    """
    获取上一个交易日 (Timestamp版本)
    """
    trade_dates = __calendar_timestamps()
    if not trade_dates:
        return Timestamp.zero()

    if date is None:
        date = Timestamp.now().get_pre_market_time()

    # 查找 date 在 trade_dates 中的位置 (lower_bound)
    # bisect_left 相当于 C++ 的 lower_bound
    idx = bisect.bisect_left(trade_dates, date)
    
    if idx > 0:
        idx -= 1
        
    # 判断是否盘前
    last_ts = trade_dates[idx]
    current_ts = debug_timestamp if debug_timestamp is not None else Timestamp.now()
    
    if current_ts < last_ts and idx > 0:
        idx -= 1
        
    return trade_dates[idx]


def next_trading_day_ts(date: Optional[Timestamp] = None, debug_timestamp: Optional[Timestamp] = None) -> Timestamp:
    """
    获取下一个交易日 (Timestamp版本)
    """
    trade_dates = __calendar_timestamps()
    if not trade_dates:
        return Timestamp.zero()

    if date is None:
        date = Timestamp.now().get_pre_market_time()
        
    current_time = debug_timestamp if debug_timestamp is not None else Timestamp.now()
    
    # 找到第一个大于等于 date 的交易日 (lower_bound)
    idx = bisect.bisect_left(trade_dates, date)
    
    if idx >= len(trade_dates):
        if trade_dates:
            return trade_dates[-1]
        return Timestamp.zero()
        
    candidate_day = trade_dates[idx]
    
    # 如果当前时间已经过了候选交易日的盘前时间，则取下一个
    if current_time >= candidate_day and idx < len(trade_dates):
        idx += 1
        if idx >= len(trade_dates):
            return trade_dates[-1]
        return trade_dates[idx]
        
    return candidate_day


def date_range(begin: Timestamp, end: Optional[Timestamp] = None, skip_today: bool = False) -> List[Timestamp]:
    """
    获取日期范围 (Timestamp版本)
    """
    if end is None:
        end = Timestamp.now()
        
    if begin > end:
        return []
        
    trade_dates = __calendar_timestamps()
    if not trade_dates:
        return []
        
    # 查找范围边界
    lower = bisect.bisect_left(trade_dates, begin)
    upper = bisect.bisect_right(trade_dates, end)
    
    # 处理 skip_today 逻辑
    if skip_today and upper > 0:
        today = Timestamp.now().get_pre_market_time()
        last_in_range = trade_dates[upper - 1]
        if last_in_range > today or last_in_range > end:
            upper -= 1
    else:
        # 调整 upper 到最后一个 <= end 的日期
        # bisect_right 返回的是第一个 > end 的位置，所以 upper-1 就是 <= end 的位置
        # 这里不需要额外调整，切片 [lower:upper] 刚好包含 lower 到 upper-1
        pass
        
    if lower >= upper:
        return []
        
    return trade_dates[lower:upper]


def get_date_range(begin: str, end: str, skip_today: bool = False) -> List[str]:
    """
    获取日期范围 (字符串版本)
    """
    if begin > end:
        return []
        
    trade_dates = load_calendar().tolist() # Convert Series to list for bisect
    if not trade_dates:
        return []
        
    # 查找起始索引
    it_start = bisect.bisect_left(trade_dates, begin)
    
    # 查找结束索引
    it_end = bisect.bisect_left(trade_dates, end)
    
    # bisect_left 返回的是第一个 >= end 的位置
    # 如果 trade_dates[it_end] == end, 我们需要包含它，所以切片应该是 it_end + 1
    # 如果 trade_dates[it_end] > end, 我们不需要包含它，切片应该是 it_end
    
    # 为了匹配 C++ lower_bound 逻辑:
    # C++: lower_bound(begin), lower_bound(end)
    # range: [itStart, itEnd) if *itEnd >= end? No, C++ logic is complex.
    # Let's simplify: find all dates d such that begin <= d <= end
    
    # Re-implement using simple filtering or bisect_right for end
    
    it_start = bisect.bisect_left(trade_dates, begin)
    it_end = bisect.bisect_right(trade_dates, end) # first element > end
    
    if skip_today:
        if it_end > 0:
            today_str = get_today()
            last_day = trade_dates[it_end - 1]
            if last_day > today_str or last_day > end:
                it_end -= 1
                
    if it_start >= it_end:
        return []
        
    return trade_dates[it_start:it_end]


if __name__ == '__main__':
    a1 = last_trading_day()
    print(type(a1))
    print(a1)

File: quant1x/data/exchange/calendar.go
package exchange

import (
	"encoding/csv"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"gitee.com/quant1x/quant1x/quant1x/core"
	"gitee.com/quant1x/quant1x/quant1x/exchange/sina"
	"gitee.com/quant1x/quant1x/quant1x/runtime"

	"time"
)

// CalendarFilename returns full path to calendar cache file under meta directory.
// Previously this helper lived in package `meta`; migrated here to keep
// exchange-specific cache layout nearby.
func CalendarFilename() string {
	return filepath.Join(core.GetMetaPath(), "calendar")
}

// calendarMarkerFilename returns the path to the marker file used to record
// the last successful calendar update. Kept private to this package.
func calendarMarkerFilename() string {
	return filepath.Join(core.GetMetaPath(), "calendar.updated")
}

const (
	sinaCalendarURL     = "https://finance.sina.com.cn/realstock/company/klc_td_sh.txt"
	calendarMissingDate = "1992-05-04"
)

var (
	// in-memory caches
	globalCalendarsString    []string
	globalCalendarsTimestamp []Timestamp
	// Initialize a default RollingOnce using the central `runtime` package so call sites can invoke it directly
	// and behavior matches C++'s global_calendar_once->Do(...).
	calendarRollingOnce = runtime.RollingOnceDaily(PreMarketHour, PreMarketMinute)
)

// preprocess JS-like response text (strip assignment, trailing semicolon, and quotes)
// Name aligned with C++ `preprocess` for 1:1 parity.
func preprocess(text string) string {
	s := text
	if pos := strings.Index(s, "="); pos != -1 {
		s = s[pos+1:]
	}
	if pos := strings.Index(s, ";"); pos != -1 {
		s = s[:pos]
	}
	s = strings.ReplaceAll(s, "\"", "")
	return s
}

// decode mirrors the C++ detail::decode: preprocess then use the sina finance
// decoder and extract the "date" field from decoded records preserving order.
func decode(text string) []string {
	pre := preprocess(text)
	dec := sina.NewFinanceDecoder(pre)
	raw := dec.Decode()

	dates := make([]string, 0)
	switch v := raw.(type) {
	case []map[string]string:
		for _, m := range v {
			if d, ok := m["date"]; ok && strings.TrimSpace(d) != "" {
				dates = append(dates, d)
			}
		}
	case []map[string]any:
		for _, m := range v {
			if ai, ok := m["date"]; ok {
				if ds, ok2 := ai.(string); ok2 && strings.TrimSpace(ds) != "" {
					dates = append(dates, ds)
				}
			}
		}
	case []string:
		dates = append(dates, v...)
	default:
		// no-op, return empty/nil to indicate nothing decoded
	}

	if len(dates) == 0 {
		return nil
	}
	return dates
}

// updateCalendar downloads calendar data from Sina and caches it to disk.
func updateCalendar() error {
	fname := CalendarFilename()
	// attempt conditional GET if file exists
	var modtime time.Time
	if fi, err := os.Stat(fname); err == nil {
		modtime = fi.ModTime()
	}

	client := &http.Client{Timeout: 15 * time.Second}
	req, err := http.NewRequest(http.MethodGet, sinaCalendarURL, nil)
	if err != nil {
		return err
	}
	if !modtime.IsZero() {
		req.Header.Set("If-Modified-Since", modtime.UTC().Format(http.TimeFormat))
	}
	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusNotModified {
		return nil
	}
	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("http status %d", resp.StatusCode)
	}

	bodyBytes, err := io.ReadAll(resp.Body)
	if err != nil {
		return err
	}
	body := string(bodyBytes)

	// decode using the package-local helper which mirrors C++ detail::decode
	dates := decode(body)

	// ensure missing date present
	// Insert missing date in-place (do NOT re-sort) to preserve decoder order,
	// matching the C++ behavior which does not explicitly sort the decoded list.
	idx := sort.SearchStrings(dates, calendarMissingDate)
	if idx == len(dates) || dates[idx] != calendarMissingDate {
		// insert at idx
		dates = append(dates, "")
		copy(dates[idx+1:], dates[idx:])
		dates[idx] = calendarMissingDate
	}

	// write CSV cache
	if err := os.MkdirAll(filepath.Dir(fname), 0o755); err != nil {
		return err
	}
	f, err := os.Create(fname)
	if err != nil {
		return err
	}
	defer f.Close()
	w := csv.NewWriter(f)
	_ = w.Write([]string{"date", "source"})
	for _, d := range dates {
		_ = w.Write([]string{d, "sina"})
	}
	w.Flush()
	if err := w.Error(); err != nil {
		return err
	}

	// set file mtime if Last-Modified provided
	if lm := resp.Header.Get("Last-Modified"); lm != "" {
		if t, e := http.ParseTime(lm); e == nil {
			_ = os.Chtimes(fname, t, t)
		}
	}

	// create/update marker file next to calendar cache so callers can
	// decide whether a remote update is necessary without hitting the
	// network. Use the local completion time (now) as the marker mtime —
	// the marker indicates when we last successfully updated the cache,
	// not when the remote resource claims to have been last-modified.
	marker := calendarMarkerFilename()
	markerTime := time.Now()
	if mf, err := os.Create(marker); err == nil {
		_ = mf.Close()
		_ = os.Chtimes(marker, markerTime, markerTime)
	}
	return nil
}

// lazyLoadCalendar ensures the calendar cache exists and loads it into memory.
func lazyLoadCalendar() {
	fname := CalendarFilename()
	// marker file lives next to calendar cache
	marker := calendarMarkerFilename()

	// Determine today's pre-market timestamp via session.GetTodayInit()
	tsTodayInit := GetTodayInit()
	//nowTs := NowTimestamp()

	ensureUpdated := false
	tsMarkerModified, err := GetFilenameModifiedTime(marker)
	if err != nil {
		// marker missing: ensure update
		ensureUpdated = true
	} else {
		// if marker modified time < today's pre-market time, ensure update
		if tsMarkerModified.Less(tsTodayInit) {
			ensureUpdated = true
		}
	}

	if ensureUpdated {
		_ = updateCalendar()
	}
	f, err := os.Open(fname)
	if err != nil {
		return
	}
	defer f.Close()
	r := csv.NewReader(f)
	// read header
	if _, err := r.Read(); err != nil {
		return
	}
	var ss []string
	var ts []Timestamp
	for {
		rec, err := r.Read()
		if err == io.EOF {
			break
		}
		if err != nil || len(rec) == 0 {
			continue
		}
		date := strings.TrimSpace(rec[0])
		if date == "" {
			continue
		}
		ss = append(ss, date)
		if t, e := ParseTimestamp(date); e == nil {
			ts = append(ts, t.PreMarketTime())
		}
	}
	globalCalendarsString = ss
	globalCalendarsTimestamp = ts
}

// get_calendar_list returns the in-memory calendar (loads on first call).
// Name aligned with C++ `get_calendar_list` for 1:1 parity.
func get_calendar_list() []string {
	// Ensure calendar is loaded; use persistent RollingOnce (parity with C++ global_calendar_once).
	calendarRollingOnce.Do(lazyLoadCalendar)
	if len(globalCalendarsString) == 0 {
		panic("exchange calendar is empty")
	}
	return append([]string(nil), globalCalendarsString...)
}

func get_date_range(begin, end string, skipToday bool) []string {
	if begin > end {
		return nil
	}
	list := get_calendar_list()
	is := sort.SearchStrings(list, begin)
	ie := sort.SearchStrings(list, end)

	// Follow C++ ordering: handle skipToday first, else adjust ie to <= end
	if skipToday {
		if ie < len(list) {
			today := NowTimestamp().OnlyDate()
			lastDay := list[ie]
			if lastDay > today || lastDay > end {
				ie--
			}
		}
	} else {
		for ie >= 0 && ie < len(list) && list[ie] > end {
			ie--
		}
	}

	if is < 0 || ie < 0 || is > ie || ie >= len(list) {
		return nil
	}
	return append([]string(nil), list[is:ie+1]...)
}

func DateRange(begin, end Timestamp, skipToday bool) []Timestamp {
	// Mirror C++: operate directly on the timestamp cache (pre-market times)
	calendarRollingOnce.Do(lazyLoadCalendar)
	tradeDates := globalCalendarsTimestamp
	if len(tradeDates) == 0 {
		return nil
	}

	n := len(tradeDates)

	// lower_bound: first >= begin
	lower := sort.Search(n, func(i int) bool { return !tradeDates[i].Less(begin) })

	// upper_bound: first > end
	upper := sort.Search(n, func(i int) bool { return tradeDates[i].Greater(end) })

	if skipToday && upper < n {
		today := NowTimestamp().PreMarketTime()
		// if candidate at upper is > today or > end, step back
		if tradeDates[upper].Greater(today) || tradeDates[upper].Greater(end) {
			if upper > 0 {
				upper--
			}
		}
	} else {
		// adjust upper to the last index <= end
		for upper > 0 && tradeDates[upper-1].Greater(end) {
			upper--
		}
	}

	// validity checks (mirror C++ conditions)
	if lower >= upper || lower == n || upper == 0 {
		return nil
	}

	// return half-open [lower, upper)
	out := make([]Timestamp, upper-lower)
	copy(out, tradeDates[lower:upper])
	return out
}

// LastTradingDay returns the most recent trading day <= date
func LastTradingDay(t Timestamp) Timestamp {
	// Ensure calendar is loaded (persistent RollingOnce, parity with C++ global_calendar_once).
	calendarRollingOnce.Do(lazyLoadCalendar)
	if len(globalCalendarsTimestamp) == 0 {
		return NowTimestamp().PreMarketTime()
	}
	// Follow C++ logic: find upper_bound (first > t), then step back to <= t.
	n := len(globalCalendarsTimestamp)
	i := sort.Search(n, func(i int) bool { return globalCalendarsTimestamp[i].Greater(t) })
	if i != 0 {
		i-- // now globalCalendarsTimestamp[i] <= t
	}

	lastTimestamp := globalCalendarsTimestamp[i]
	currentTimestamp := NowTimestamp()
	// If current time is before lastTimestamp (盘前), move to previous trading day if possible
	if currentTimestamp.Less(lastTimestamp) && i != 0 {
		i--
	}
	return globalCalendarsTimestamp[i]
}
